### PR TITLE
neorpc: adjust `SignerWithWitness` marshalling scheme

### LINKED ISSUE #3059: discrepancy in parsing witness scope format on RPC side
The RPC method `invokescript`  supports passing optional `signers`. While investigating a wallet error reported by a user @hal0x2328 noticed that the Go node fails to parse a signer provided whereas the C# node doesn't.

Example request made;
```
{"params":["CxEMFM8qjpTLqLqEOKINSBGOosvj8jVKDBTPKo6Uy6i6hDiiDUgRjqLL4/I1ShTAHwwIdHJhbnNmZXIMFPVj6kC8KD1NDgXEjqMFs/Kgc0DvQWJ9W1I=",[{"account":"4a35f2e3cba28e11480da23884baa8cb948e2acf","scopes":1}]],"jsonrpc":"2.0","id":1234,"method":"invokescript"}
```

Go node response
```
{"id":1234,"jsonrpc":"2.0","error":{"code":-32602,"message":"Invalid Params","data":"not a signer: not a signer: json: cannot unmarshal number into Go struct field signerWithWitnessAux.scopes of type string"}}
```

Note how `scopes` is defined as a number not as a name of the flag, which C# successfully parses [here](https://github.com/neo-project/neo-modules/blob/b639e84c8fa7aa33c063ba764b27b9f93ef0ef04/src/RpcServer/RpcServer.SmartContract.cs#L171)

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #3059. C#'s rules for RPC-server signers unmarshalling are not as strict as for core signers. Not sure whether it's worth bug in C# node, because it's a valid behaviour. Let's fix it in our side.

### CODE REVIEW

**Review by cthulhu-rider — APPROVED** [review]

File: pkg/core/transaction/witness_scope.go [hunk 1]
```diff
@@ -29,6 +29,19 @@ const (
 	Global WitnessScope = 0x80
 )
 
+// ScopesFromByte converts byte to a set of WitnessScopes and performs validity
+// check.
+func ScopesFromByte(b byte) (WitnessScope, error) {
+	var res = WitnessScope(b)
+	if (res&Global != 0) && (res&(None|CalledByEntry|CustomContracts|CustomGroups|Rules) != 0) {
+		return 0, fmt.Errorf("Global scope can not be combined with other scopes")
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/core/transaction/witness_scope.go]: > mb errors.New/errvar is enough

Good, it's a mistake from my side, fixed.

> double-check

Yep, it's entantionally, that's the scope name, it's not "global".

File: pkg/neorpc/types.go [hunk 2]
```diff
@@ -92,9 +92,13 @@ type signerWithWitnessAux struct {
 
 // MarshalJSON implements the json.Marshaler interface.
 func (s *SignerWithWitness) MarshalJSON() ([]byte, error) {
+	sc, err := s.Scopes.MarshalJSON()
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal scopes: %w", err)
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/neorpc/types.go:97]: We always have upper-level wrapper, so it's clear.

File: pkg/services/rpcsrv/params/param_test.go [hunk 3]
```diff
@@ -264,16 +264,40 @@ func TestGetWitness(t *testing.T) {
...
+				Scopes:  transaction.CalledByEntry,
+			}},
+		},
+		{`{"account": "0xcadb3dc2faa3ef14a13b619c9a43124755aa2569", "scopes": 17}`, neorpc.SignerWithWitness{
+			Signer: transaction.Signer{
+				Account: accountHash,
+				Scopes:  transaction.CalledByEntry | transaction.CustomContracts,
+			}},
+		},
 	}
```
Comment (roman-khimov) [inline pkg/services/rpcsrv/params/param_test.go:292]: How about bad cases?

**Review by roman-khimov — COMMENTED** [review]
I'd expect to see `WitnessScope.UnmarshalJSON` modification only, but if it's really unmarshaled differently in different contexts then I love C#, yay.

**Comment by AnnaShaleva** [comment]
> I'd expect to see `WitnessScope.UnmarshalJSON` modification only, but if it's really unmarshaled differently in different contexts then I love C#, yay.

I've even added a unit-test to C# node to ensure that witness scope of a simple signer can't be unmarshalled from number (only from enum's string representation).

File: pkg/services/rpcsrv/params/param_test.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/services/rpcsrv/params/param_test.go:322]: Fixed.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**